### PR TITLE
fixed incorrect WHERE caused by type as primary key in STI

### PR DESCRIPTION
**Pull request by aolvictim**
This is a potential fix for the issue https://github.com/typeorm/typeorm/issues/4252

Basically it checks if the discriminator value is present in the entity, if not, add it.

### CODE REVIEW

**Comment by YegorZaremba** [comment]
It'll be better if you write a tests for this case

**Comment by pleerock** [comment]
First of all we need to narrow down issue scope. Is this issue only when you use STI? Can we confirm it by having a test for this case (your issue without STI, the most simple case). If issue exist then this fix isn't correct. The reason I ask is because we need to fix a source of problem, which is hard to find in this case.

**Review by imnotjames — CHANGES_REQUESTED** [review]
This needs tests to confirm that the behavior you expect from this code is exhibited.

File: src/metadata/EntityMetadata.ts [hunk 1]
```diff
@@ -570,7 +570,11 @@ export class EntityMetadata {
         if (!entity)
             return undefined;
 
-        return EntityMetadata.getValueMap(entity, this.primaryColumns, { skipNulls: true });
+        let fullEntity = entity;
+        if (this.discriminatorColumn && !entity.hasOwnProperty(this.discriminatorColumn.databasePath))
+            fullEntity[this.discriminatorColumn.databasePath] = this.discriminatorValue;
```
Comment (imnotjames) [inline src/metadata/EntityMetadata.ts:575]: Not sure I understand why you've assigned this variable and then are changing the prop in this variable.  Could you explain it to me?

If it's an object it's the same object in both variables - you haven't cloned it - so you might as well just `entity[this.discriminatorColumn.databasePath] = this.discriminatorValue;`

**Comment by AlexMesser** [comment]
Closing as there is no answer from OP.

**Pull request closed without merging**